File: HashMap/1. Two Sum 6 Feb 2022/twoSum.py
# O(n)
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # O(n)
        # if nums[i] is the same, i will be the last index
        # [2,5,5,11]
        # {2: 0, 5: 2, 11: 3}
        numsDictionary = {nums[i] : i for i in range(len(nums))}
        
        
        # O(n)
        for i, num in enumerate(nums):
            diff = target - num
            
            # numsDictionary[diff] might not exist
            # e.g. 10 - 2 = 8 results in key error
            # if statement must check if diff is in the dictionary first
            # if not, short circuit and iterate to the next value
            if diff in numsDictionary and i != numsDictionary[diff]:
                return [i, numsDictionary[diff]]
        
            
            
# O(n^2)
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        for i in range(len(nums)):
            current = nums[i]
            for j in range(i + 1, len(nums)):        
                if nums[i] + nums[j] == target:
                    return [i, j]
            

File: HashMap/1. Two Sum 6 Feb 2022/twoSum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {
    let map = new Map();
    for (let i = 0; i < nums.length; i++) {
        map.set(nums[i], i);
    }

    for (let i = 0; i < nums.length; i++) {
        let diff = target - nums[i];
        if (map.has(diff) && map.get(diff) !== i) {
            return [map.get(diff), i];
        }
    }
};



/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
//Brute Force
var twoSum = function (nums, target) {
    let result = [];

    for (let i = 0; i < nums.length; i++) {
        let diff = target - nums[i];
        for (let j = i + 1; j < nums.length; j++) {
            if (nums[j] === diff) {
                result.push(i);
                result.push(j);
            }
        }
    }

    return result;
};